polygeist/tools/cgeist/Lib/Attributes.h

```diff
@@ -9,90 +9,216 @@
 #ifndef CGEIST_ATTRIBUTES_H
 #define CGEIST_ATTRIBUTES_H
 
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
 #include "mlir/IR/Attributes.h"
 #include "mlir/IR/OperationSupport.h"
 #include "llvm/IR/Attributes.h"
 
 namespace mlirclang {
 
+class AttrBuilder;
+
+/// \class
+/// This class holds the attributes for a function, its return value, and
+/// its parameters. You access the attributes for each of them via an index into
+/// the AttributeList object. The function attributes are at index
+/// `AttributeList::FunctionIndex', the return value is at index
+/// `AttributeList::ReturnIndex', and the attributes for the parameters start at
+/// index `AttributeList::FirstArgIndex'.
+class AttributeList {
+public:
+  AttributeList() = default;
+
+  //===--------------------------------------------------------------------===//
+  // AttributeList Mutation
+  //===--------------------------------------------------------------------===//
+
+  /// Add function. return value, and parameters attributes to the list.
+  AttributeList &addAttrs(const AttrBuilder &FnAttrB,
+                          const AttrBuilder &RetAttrB,
+                          llvm::ArrayRef<mlir::NamedAttrList> Attrs);
+
+  /// Add function attributes to the list.
+  AttributeList &addFnAttrs(const AttrBuilder &B);
+
+  /// Add return value attributes to the list.
+  AttributeList &addRetAttrs(const AttrBuilder &B);
+
+  /// Add parameters attributes to the list.
+  AttributeList &addParmAttrs(llvm::ArrayRef<mlir::NamedAttrList> Attrs);
+
+  /// The function attributes are returned.
+  mlir::NamedAttrList getFnAttrs() const { return FnAttrs; }
+
+  /// The attributes for the ret value are returned.
+  mlir::NamedAttrList getRetAttrs() const { return RetAttrs; }
+
+  /// The attributes for the parameters are returned.
+  mlir::ArrayRef<mlir::NamedAttrList> getParmAttrs() const { return ParmAttrs; }
```

Comment:
Do you think we can drop this and use the function below instead? When will this be used?